### ISSUE TITLE
On Make - check minimal rebar3 version

### ISSUE THREAD

**Issue by mworrell** [zotonic/zotonic#1876]
Sometimes we use new capabilities of rebar3.
But an older version of rebar3 is not replaced.

We should check the installed version of rebar3, and if needed replace it with a newer version.

See also #1875

**Comment by karlsson** [comment]
I checked this a bit and one can use the HTTP HEAD request to get the ETag from Amazon and check if this is changed to decide if to install a new version of rebar3.
This works fine even though the erlang eval string gets a bit long. If one expect to always have curl at hand maybe this is an alternative, but I do not know how this works on other platforms than Linux.
Also if rebar3 is updated with a non-backwards compatibility you may be cooked (like an oliebol :-) ).
If version control needs to be really hard-coded maybe one should check if it is possible to set a rebar3 version as a dependency in rebar.config itself, get it and make it from there (using rebar3 to make rebar3...).

From the GNUmakefile:
```# Erlang Rebar downloading
# see: https://groups.google.com/forum/?fromgroups=#!topic/erlang-programming/U0JJ3SeUv5Y
REBAR := ./rebar3
REBAR_URL := https://s3.amazonaws.com/rebar3/rebar3
REBAR_ETAG := ./rebar_etag

REBAR_OPTS ?=
# Default target - update sources and call all compile rules in succession
all: compile
.PHONY: all

# Get ETAG from rebar3 amazon cloud and update etag file if different.
get_rebar_tag:
	$(ERL) -noshell -s inets -s ssl \
	  -eval 'case httpc:request(head, {"$(REBAR_URL)", []}, [{timeout, 2000}], []) of {ok, {_, Headers,_}} -> Etag = proplists:get_value("etag", Headers), Bin = list_to_binary(Etag), case file:read_file("$(REBAR_ETAG)") of {ok, Bin} -> io:fwrite("ETag update not needed~n"); {ok, _OldEtag} -> file:write_file("$(REBAR_ETAG)", Bin); {error, enoent} -> file:write_file("$(REBAR_ETAG)", Bin); _ -> ok end,   io:fwrite("Etag: ~s~n",[Etag]); Error -> io:fwrite("Failed to get rebar3 etag: ~p~n",[Error]) end' \
	  -s init stop

$(REBAR_ETAG): get_rebar_tag

$(REBAR): $(REBAR_ETAG)
	$(ERL) -noshell -s inets -s ssl \
	  -eval 'file:delete("$(REBAR)"),{ok, saved_to_file} = httpc:request(get, {"$(REBAR_URL)", []}, [], [{stream, "$(REBAR)"}])' \
	  -s init stop
	chmod +x $(REBAR)```

**Comment by mworrell** [comment]
That looks complicated indeed.  Good work to figure that out.

I was thinking about your remark about the "oliebol" situation. That is indeed a concern, that your application changes whilst rebuilding it because rebar3 changed.

Of course, this whole issue was about people having ancient rebar3 installs in their directory which don't support newer functionalities we need. That will be nicely fixed with your changes.

And we always download rebar3 and use it on Travis-CI anyway, so if there is a wrong one we will see it quite soon. And it will fix itself, using your update method.

Would be nice if rebar3 had published release versions which we can grab.

I think what you made solves a bigger issue than the potential problem of being cooked.

What is your opinion?

**Comment by karlsson** [comment]
Yes, I guess that folks download the latest rebar3 manually from Amazon and would be "cooked" anyway in case that rebar3 version gets incompatible and as long as development is fairly active the build chain will be checked and up to date so pulling the latest in the Makefile is better than just installing once and never updating.
Question is if you like to include such an Erlang code in the Makefile as in the pull request or have some (esthetical or other) concerns.

**Comment by karlsson** [comment]
Also, I assume that s3.amazonaws.com behaves properly and updates the ETag whenever a new rebar3 version is pushed. This I cannot test myself.

**Comment by mworrell** [comment]
I think the ETag assumption is right. The ETag should have some value that changes if the resource changes.
